src/digest/sha1.rs

```diff
@@ -49,38 +49,46 @@ fn block_data_order_safe(state: &mut [u64; MAX_CHAINING_LEN / 8],
     let state = &mut state[..CHAINING_WORDS];
     let state = slice_as_array_ref_mut!(state, CHAINING_WORDS).unwrap();
 
-    let mut w: [W32; 80] = [Wrapping(0); 80];
+    let mut w: [W32; 16] = [Wrapping(0); 16];
     for block in blocks {
-        for t in 0..16 {
-            let word = slice_as_array_ref!(&block[t * 4..][..4], 4).unwrap();
-            w[t] = Wrapping(polyfill::slice::u32_from_be_u8(word))
-        }
-        for t in 16..80 {
-            let wt = w[t - 3] ^ w[t - 8] ^ w[t - 14] ^ w[t - 16];
-            w[t] = polyfill::wrapping_rotate_left_u32(wt, 1);
-        }
-
         let mut a = state[0];
         let mut b = state[1];
         let mut c = state[2];
         let mut d = state[3];
         let mut e = state[4];
 
-        for t in 0..80 {
+        macro_rules! round {
+            ($k: expr, $f: expr, $w: expr) => {
+                {
+                    let tt = polyfill::wrapping_rotate_left_u32(a, 5) + $f + e +
+                             Wrapping($k) + $w;
```

Comment:
It may be worth asking the rustc people for help. In particular, do they think there's any hope in the compiler auto-vectoring this code? If not, what would need to be changed for it to be auto-vectorizable? If auto-vectorization is impossible, is there a platform-neutral explicit vectorization we should use?

I suspect that in order for this to be vectorized, you'd need to target a platform that has PSHUFB or equivalent, and you may need to rewrite the rotations and shifts.

It would definitely be interesting if it can be done in a way that is a performance win. But, I don't htink it makes sense to make things more complicated and *also* make them slower.